File: julia/src/GmeowGTS.jl
# SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: MIT OR Apache-2.0

module GmeowGTS

using Libdl

export ABI_VERSION,
    GtsError,
    Status,
    UnpackFlags,
    abi_version,
    build_metadata_json,
    capabilities_json,
    files_diff_json,
    files_pack,
    files_unpack,
    from_nquads,
    libgts_version,
    read_json,
    status_name,
    to_nquads,
    verify_json

const ABI_VERSION = UInt32(1)
const VERSION = v"0.9.4"

module Status
const OK = Int32(0)
const INVALID_ARGUMENT = Int32(1)
const IO = Int32(2)
const PARSE = Int32(3)
const DIAGNOSTIC = Int32(4)
const INTERNAL = Int32(5)
const PANIC = Int32(6)
end

module UnpackFlags
const NONE = UInt32(0)
const INCLUDE_SUPPRESSED = UInt32(1) << 0
const ALLOW_SYMLINKS = UInt32(1) << 1
const ALLOW_SPECIAL = UInt32(1) << 2
const SAME_OWNER = UInt32(1) << 3
const PRESERVE_SETID = UInt32(1) << 4
end

const STATUS_NAMES = Dict(
    Status.OK => "OK",
    Status.INVALID_ARGUMENT => "INVALID_ARGUMENT",
    Status.IO => "IO",
    Status.PARSE => "PARSE",
    Status.DIAGNOSTIC => "DIAGNOSTIC",
    Status.INTERNAL => "INTERNAL",
    Status.PANIC => "PANIC",
)

const LIBRARY_HANDLES = Dict{String,Ptr{Cvoid}}()
const LIBRARY_HANDLE_LOCK = ReentrantLock()

function library_handle(library::AbstractString)
    key = String(library)
    lock(LIBRARY_HANDLE_LOCK)
    try
        get!(LIBRARY_HANDLES, key) do
            Libdl.dlopen(key)
        end
    finally
        unlock(LIBRARY_HANDLE_LOCK)
    end
end

native_symbol(library::AbstractString, name::Symbol) = Libdl.dlsym(library_handle(library), name)

struct GtsBuffer
    data::Ptr{UInt8}
    len::Csize_t
    capacity::Csize_t
end

struct GtsError <: Exception
    operation::String
    status::Int32
    status_name::String
    code::String
    detail::String
end

function Base.showerror(io::IO, error::GtsError)
    message = "$(error.operation) failed with $(error.status_name)"
    if !isempty(error.code)
        message = "$message ($(error.code))"
    end
    if !isempty(error.detail)
        message = "$message: $(error.detail)"
    end
    print(io, message)
end

status_name(code::Integer) = get(STATUS_NAMES, Int32(code), "UNKNOWN")

function default_library()
    explicit = get(ENV, "GTS_LIBGTS", "")
    if !isempty(explicit)
        return explicit
    end

    directory = get(ENV, "GTS_LIB_DIR", "")
    library = if Sys.iswindows()
        "gts.dll"
    elseif Sys.isapple()
        "libgts.dylib"
    else
        "libgts.so"
    end

    isempty(directory) ? library : joinpath(directory, library)
end

function abi_version(; library::AbstractString = default_library())
    ccall(native_symbol(library, :gts_abi_version), UInt32, ())
end

function libgts_version(; library::AbstractString = default_library())
    version_pointer = ccall(native_symbol(library, :gts_version), Cstring, ())
    version_pointer == C_NULL ? "" : unsafe_string(version_pointer)
end

function build_metadata_json(; library::AbstractString = default_library())
    text_call("gts_build_metadata_json", library) do out, error
        ccall(native_symbol(library, :gts_build_metadata_json), Cint, (Ref{GtsBuffer}, Ref{Ptr{Cvoid}}), out, error)
    end
end

function capabilities_json(; library::AbstractString = default_library())
    text_call("gts_capabilities_json", library) do out, error
        ccall(native_symbol(library, :gts_capabilities_json), Cint, (Ref{GtsBuffer}, Ref{Ptr{Cvoid}}), out, error)
    end
end

function read_json(data; library::AbstractString = default_library())
    with_bytes(data, "data") do data_pointer, len
        text_call("gts_read_json", library) do out, error
            ccall(
                native_symbol(library, :gts_read_json),
                Cint,
                (Ptr{UInt8}, Csize_t, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                data_pointer,
                len,
                out,
                error,
            )
        end
    end
end

function verify_json(data; library::AbstractString = default_library())
    with_bytes(data, "data") do data_pointer, len
        text_call("gts_verify_json", library) do out, error
            ccall(
                native_symbol(library, :gts_verify_json),
                Cint,
                (Ptr{UInt8}, Csize_t, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                data_pointer,
                len,
                out,
                error,
            )
        end
    end
end

function to_nquads(data; library::AbstractString = default_library())
    with_bytes(data, "data") do data_pointer, len
        text_call("gts_to_nquads", library) do out, error
            ccall(
                native_symbol(library, :gts_to_nquads),
                Cint,
                (Ptr{UInt8}, Csize_t, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                data_pointer,
                len,
                out,
                error,
            )
        end
    end
end

function from_nquads(text::AbstractString; library::AbstractString = default_library())
    text_string = checked_string(text, "text")
    ensure_no_nul(text_string, "text")

    GC.@preserve text_string begin
        text_pointer = if isempty(text_string)
            Ptr{UInt8}(C_NULL)
        else
            Base.unsafe_convert(Ptr{UInt8}, Base.pointer(text_string))
        end
        raw_call("gts_from_nquads", library) do out, error
            ccall(
                native_symbol(library, :gts_from_nquads),
                Cint,
                (Ptr{UInt8}, Csize_t, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                text_pointer,
                Csize_t(sizeof(text_string)),
                out,
                error,
            )
        end
    end
end

function files_pack(paths; library::AbstractString = default_library())
    strings = checked_strings(paths, "paths")
    cstrings = Vector{Cstring}(undef, length(strings))
    for (index, path) in pairs(strings)
        cstrings[index] = checked_cstring(path, "paths[$index]")
    end

    GC.@preserve strings cstrings begin
        raw_call("gts_files_pack", library) do out, error
            ccall(
                native_symbol(library, :gts_files_pack),
                Cint,
                (Ptr{Cstring}, Csize_t, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                Base.pointer(cstrings),
                Csize_t(length(cstrings)),
                out,
                error,
            )
        end
    end
end

function files_unpack(data, destination::AbstractString; flags = UnpackFlags.NONE, library::AbstractString = default_library())
    destination_string = checked_string(destination, "destination")
    ensure_no_nul(destination_string, "destination")
    unpack_flags = checked_uint32(flags, "flags")

    GC.@preserve destination_string begin
        destination_pointer = checked_cstring(destination_string, "destination")
        with_bytes(data, "data") do data_pointer, len
            text_call("gts_files_unpack", library) do out, error
                ccall(
                    native_symbol(library, :gts_files_unpack),
                    Cint,
                    (Ptr{UInt8}, Csize_t, Cstring, UInt32, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                    data_pointer,
                    len,
                    destination_pointer,
                    unpack_flags,
                    out,
                    error,
                )
            end
        end
    end
end

function files_diff_json(data, directory::AbstractString; library::AbstractString = default_library())
    directory_string = checked_string(directory, "directory")
    ensure_no_nul(directory_string, "directory")

    GC.@preserve directory_string begin
        directory_pointer = checked_cstring(directory_string, "directory")
        with_bytes(data, "data") do data_pointer, len
            text_call("gts_files_diff_json", library) do out, error
                ccall(
                    native_symbol(library, :gts_files_diff_json),
                    Cint,
                    (Ptr{UInt8}, Csize_t, Cstring, Ref{GtsBuffer}, Ref{Ptr{Cvoid}}),
                    data_pointer,
                    len,
                    directory_pointer,
                    out,
                    error,
                )
            end
        end
    end
end

function text_call(f, operation::AbstractString, library::AbstractString)
    String(raw_buffer_call(operation, library, f))
end

function raw_call(f, operation::AbstractString, library::AbstractString)
    raw_buffer_call(operation, library, f)
end

function raw_buffer_call(operation::AbstractString, library::AbstractString, f)
    buffer_free = native_symbol(library, :gts_buffer_free)
    out = Ref(GtsBuffer(Ptr{UInt8}(C_NULL), Csize_t(0), Csize_t(0)))
    error = Ref{Ptr{Cvoid}}(C_NULL)

    try
        status = Int32(f(out, error))
        error_pointer = error[]
        if status != Status.OK
            throw(build_error(operation, status, error_pointer, library))
        end
        if error_pointer != C_NULL
            throw(build_error(operation, Status.INTERNAL, error_pointer, library))
        end
        copy_buffer(out[])
    finally
        ccall(buffer_free, Cvoid, (Ref{GtsBuffer},), out)
    end
end

function build_error(operation::AbstractString, status::Integer, error_pointer::Ptr{Cvoid}, library::AbstractString)
    code = ""
    detail = ""

    if error_pointer != C_NULL
        error_code = native_symbol(library, :gts_error_code)
        error_message = native_symbol(library, :gts_error_message)
        error_free = native_symbol(library, :gts_error_free)

        try
            code_pointer = ccall(error_code, Cstring, (Ptr{Cvoid},), error_pointer)
            detail_pointer = ccall(error_message, Cstring, (Ptr{Cvoid},), error_pointer)
            code = code_pointer == C_NULL ? "" : unsafe_string(code_pointer)
            detail = detail_pointer == C_NULL ? "" : unsafe_string(detail_pointer)
        finally
            ccall(error_free, Cvoid, (Ptr{Cvoid},), error_pointer)
        end
    end

    GtsError(String(operation), Int32(status), status_name(status), code, detail)
end

function copy_buffer(buffer::GtsBuffer)
    len = checked_length(buffer.len)
    if len == 0
        return UInt8[]
    end
    if buffer.data == C_NULL
        error("C ABI returned a null data pointer with non-zero length")
    end
    copy(unsafe_wrap(Vector{UInt8}, buffer.data, len; own = false))
end

function with_bytes(f, data, name::AbstractString)
    bytes = checked_bytes(data, name)
    GC.@preserve bytes begin
        data_pointer = isempty(bytes) ? Ptr{UInt8}(C_NULL) : Base.unsafe_convert(Ptr{UInt8}, Base.pointer(bytes))
        f(data_pointer, Csize_t(length(bytes)))
    end
end

function checked_bytes(data, name::AbstractString)
    if data isa Vector{UInt8}
        return data
    end
    if data isa AbstractVector{UInt8}
        return Vector{UInt8}(data)
    end
    throw(ArgumentError("$name must be a Vector{UInt8} or AbstractVector{UInt8}"))
end

function checked_strings(values, name::AbstractString)
    if !(values isa AbstractVector)
        throw(ArgumentError("$name must be a non-empty vector of strings"))
    end
    if isempty(values)
        throw(ArgumentError("$name must be a non-empty vector of strings"))
    end
    [checked_string(value, "$name[$index]") for (index, value) in pairs(values)]
end

function checked_string(value, name::AbstractString)
    if !(value isa AbstractString)
        throw(ArgumentError("$name must be a string"))
    end
    String(value)
end

function checked_cstring(value::String, name::AbstractString)
    ensure_no_nul(value, name)
    Base.unsafe_convert(Cstring, value)
end

function ensure_no_nul(value::AbstractString, name::AbstractString)
    if occursin('\0', value)
        throw(ArgumentError("$name must not contain NUL bytes"))
    end
end

function checked_uint32(value, name::AbstractString)
    if !(value isa Integer)
        throw(ArgumentError("$name must be an integer"))
    end
    if value < 0 || value > typemax(UInt32)
        throw(ArgumentError("$name must be a non-negative 32-bit integer"))
    end
    UInt32(value)
end

function checked_length(value::Csize_t)
    if value > typemax(Int)
        error("C ABI returned a buffer too large for this Julia runtime")
    end
    Int(value)
end

end

File: swift/Sources/GmeowGTS/GmeowGTS.swift
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

import CGts
import Foundation

#if os(Linux)
import Glibc
#else
import Darwin
#endif

public struct GtsStatus: Equatable, Hashable, Sendable, CustomStringConvertible {
    public let rawValue: UInt32

    public init(rawValue: UInt32) {
        self.rawValue = rawValue
    }

    public static let ok = GtsStatus(rawValue: 0)
    public static let invalidArgument = GtsStatus(rawValue: 1)
    public static let io = GtsStatus(rawValue: 2)
    public static let parse = GtsStatus(rawValue: 3)
    public static let diagnostic = GtsStatus(rawValue: 4)
    public static let internalError = GtsStatus(rawValue: 5)
    public static let panic = GtsStatus(rawValue: 6)

    public var name: String {
        switch self {
        case .ok:
            return "OK"
        case .invalidArgument:
            return "INVALID_ARGUMENT"
        case .io:
            return "IO"
        case .parse:
            return "PARSE"
        case .diagnostic:
            return "DIAGNOSTIC"
        case .internalError:
            return "INTERNAL"
        case .panic:
            return "PANIC"
        default:
            return "UNKNOWN"
        }
    }

    public var description: String {
        name
    }
}

public struct GtsUnpackFlags: OptionSet, Sendable {
    public let rawValue: UInt32

    public init(rawValue: UInt32) {
        self.rawValue = rawValue
    }

    public static let includeSuppressed = GtsUnpackFlags(rawValue: UInt32(GTS_UNPACK_INCLUDE_SUPPRESSED))
    public static let allowSymlinks = GtsUnpackFlags(rawValue: UInt32(GTS_UNPACK_ALLOW_SYMLINKS))
    public static let allowSpecial = GtsUnpackFlags(rawValue: UInt32(GTS_UNPACK_ALLOW_SPECIAL))
    public static let sameOwner = GtsUnpackFlags(rawValue: UInt32(GTS_UNPACK_SAME_OWNER))
    public static let preserveSetID = GtsUnpackFlags(rawValue: UInt32(GTS_UNPACK_PRESERVE_SETID))
}

public struct GtsError: Error, CustomStringConvertible {
    public let operation: String
    public let status: GtsStatus
    public let code: String
    public let detail: String

    public var description: String {
        var message = "\(operation) failed with \(status.name)"
        if !code.isEmpty {
            message += " (\(code))"
        }
        if !detail.isEmpty {
            message += ": \(detail)"
        }
        return message
    }
}

public enum GTS {
    public static var abiVersion: UInt32 {
        gts_abi_version()
    }

    public static var version: String {
        copyCString(gts_version())
    }

    public static func buildMetadataJSON() throws -> String {
        try callString("gts_build_metadata_json") { out, error in
            gts_build_metadata_json(out, error)
        }
    }

    public static func capabilitiesJSON() throws -> String {
        try callString("gts_capabilities_json") { out, error in
            gts_capabilities_json(out, error)
        }
    }

    public static func readJSON(_ data: Data) throws -> String {
        try withUnsafeBytes(data) { pointer in
            try callString("gts_read_json") { out, error in
                gts_read_json(pointer, data.count, out, error)
            }
        }
    }

    public static func verifyJSON(_ data: Data) throws -> String {
        try withUnsafeBytes(data) { pointer in
            try callString("gts_verify_json") { out, error in
                gts_verify_json(pointer, data.count, out, error)
            }
        }
    }

    public static func toNQuads(_ data: Data) throws -> String {
        try withUnsafeBytes(data) { pointer in
            try callString("gts_to_nquads") { out, error in
                gts_to_nquads(pointer, data.count, out, error)
            }
        }
    }

    public static func fromNQuads(_ text: String) throws -> Data {
        let bytes = Array(text.utf8)
        return try bytes.withUnsafeBufferPointer { buffer in
            let pointer = buffer.baseAddress.map {
                UnsafeRawPointer($0).assumingMemoryBound(to: CChar.self)
            }
            return try callData("gts_from_nquads") { out, error in
                gts_from_nquads(pointer, bytes.count, out, error)
            }
        }
    }

    public static func filesPack(paths: [String]) throws -> Data {
        guard !paths.isEmpty else {
            throw invalidArgument("gts_files_pack", "paths must not be empty")
        }

        var nativePaths: [UnsafePointer<CChar>?] = []
        nativePaths.reserveCapacity(paths.count)
        defer {
            for path in nativePaths {
                if let path {
                    free(UnsafeMutableRawPointer(mutating: path))
                }
            }
        }

        for path in paths {
            guard !path.utf8.contains(0) else {
                throw invalidArgument("gts_files_pack", "path entries must not contain NUL bytes")
            }
            guard let duplicate = strdup(path) else {
                throw GtsError(
                    operation: "gts_files_pack",
                    status: .internalError,
                    code: "gts.swift.allocation_failed",
                    detail: "Unable to allocate a native path string."
                )
            }
            nativePaths.append(UnsafePointer(duplicate))
        }

        return try nativePaths.withUnsafeBufferPointer { buffer in
            try callData("gts_files_pack") { out, error in
                gts_files_pack(buffer.baseAddress, buffer.count, out, error)
            }
        }
    }

    public static func filesUnpack(
        _ data: Data,
        to destination: String,
        flags: GtsUnpackFlags = []
    ) throws -> String {
        guard !destination.utf8.contains(0) else {
            throw invalidArgument("gts_files_unpack", "destination must not contain NUL bytes")
        }

        return try destination.withCString { destinationPointer in
            try withUnsafeBytes(data) { dataPointer in
                try callString("gts_files_unpack") { out, error in
                    gts_files_unpack(dataPointer, data.count, destinationPointer, flags.rawValue, out, error)
                }
            }
        }
    }

    public static func filesDiffJSON(_ data: Data, directory: String) throws -> String {
        guard !directory.utf8.contains(0) else {
            throw invalidArgument("gts_files_diff_json", "directory must not contain NUL bytes")
        }

        return try directory.withCString { directoryPointer in
            try withUnsafeBytes(data) { dataPointer in
                try callString("gts_files_diff_json") { out, error in
                    gts_files_diff_json(dataPointer, data.count, directoryPointer, out, error)
                }
            }
        }
    }

    private static func callString(
        _ operation: String,
        _ call: (UnsafeMutablePointer<gts_buffer>?, UnsafeMutablePointer<OpaquePointer?>?) -> gts_status
    ) throws -> String {
        let data = try callData(operation, call)
        return String(decoding: data, as: UTF8.self)
    }

    private static func callData(
        _ operation: String,
        _ call: (UnsafeMutablePointer<gts_buffer>?, UnsafeMutablePointer<OpaquePointer?>?) -> gts_status
    ) throws -> Data {
        var output = gts_buffer()
        var error: OpaquePointer?
        defer {
            gts_buffer_free(&output)
        }

        let status = GtsStatus(cStatus: call(&output, &error))
        guard status == .ok else {
            throw buildError(operation: operation, status: status, error: error)
        }
        if let error {
            throw buildError(operation: operation, status: GtsStatus.internalError, error: error)
        }
        return try copyBuffer(output, operation: operation)
    }

    private static func withUnsafeBytes<T>(
        _ data: Data,
        _ body: (UnsafePointer<UInt8>?) throws -> T
    ) throws -> T {
        try data.withUnsafeBytes { rawBuffer in
            try body(rawBuffer.bindMemory(to: UInt8.self).baseAddress)
        }
    }

    private static func copyBuffer(_ buffer: gts_buffer, operation: String) throws -> Data {
        let length = Int(buffer.len)
        if length == 0 {
            return Data()
        }
        guard let data = buffer.data else {
            throw GtsError(
                operation: operation,
                status: .internalError,
                code: "gts.swift.null_buffer",
                detail: "C ABI returned a null data pointer with non-zero length."
            )
        }
        return Data(bytes: data, count: length)
    }

    private static func copyCString(_ value: UnsafePointer<CChar>?) -> String {
        guard let value else {
            return ""
        }
        return String(cString: value)
    }

    private static func buildError(
        operation: String,
        status: GtsStatus,
        error: OpaquePointer?
    ) -> GtsError {
        var code = ""
        var detail = ""
        if let error {
            code = copyCString(gts_error_code(error))
            detail = copyCString(gts_error_message(error))
            gts_error_free(error)
        }
        return GtsError(operation: operation, status: status, code: code, detail: detail)
    }

    private static func invalidArgument(_ operation: String, _ detail: String) -> GtsError {
        GtsError(
            operation: operation,
            status: .invalidArgument,
            code: "gts.swift.invalid_argument",
            detail: detail
        )
    }
}

private extension GtsStatus {
    init(cStatus: gts_status) {
        self.init(rawValue: UInt32(cStatus.rawValue))
    }
}
